
Author: guillem (updated by rmh)
Status: debian part in BTS, upstream part being handled by Guillem

diff -ur util-linux-2.12p.old/debian/patches/00list util-linux-2.12p/debian/patches/00list
--- util-linux-2.12p.old/debian/patches/00list	2005-10-06 18:45:25.000000000 +0200
+++ util-linux-2.12p/debian/patches/00list	2005-10-06 19:47:00.000000000 +0200
@@ -1,4 +1,3 @@
-10agetty
 10cal-widechar
 10cfdisk
 10debian
diff -ur util-linux-2.12p.old/debian/rules util-linux-2.12p/debian/rules
--- util-linux-2.12p.old/debian/rules	2005-10-06 18:45:25.000000000 +0200
+++ util-linux-2.12p/debian/rules	2005-10-06 19:52:35.000000000 +0200
@@ -25,7 +25,7 @@
 sparc = $(findstring $(arch),sparc)
 nohwclock = $(findstring $(arch),s390)
 
-SUBDIRS=po lib getopt disk-utils login-utils misc-utils mount sys-utils text-utils
+SUBDIRS=po lib getopt disk-utils login-utils misc-utils sys-utils text-utils
 ifeq ($(arch),$(fdisk_arch))
 SUBDIRS += fdisk
 endif
@@ -36,6 +36,7 @@
 ifneq ($(arch),$(nohwclock))
 SUBDIRS += hwclock 
 endif
+SUBDIRS += mount
 endif
 
 ifneq ($(DEB_HOST_GNU_SYSTEM),linux-gnu)
@@ -68,9 +69,12 @@
 
 SUIDFILES = debian/tmp-mount/bin/{u,}mount
 BINFILES  = sys-utils/arch text-utils/more
-UBINFILES = sys-utils/{ipcs,ipcrm,setsid} \
-	misc-utils/{namei,setterm,mcookie,whereis,ddate} \
+UBINFILES = sys-utils/{ipcrm,setsid} \
+	misc-utils/{namei,mcookie,whereis,ddate} \
 	getopt/getopt text-utils/{rev,line,pg}
+ifeq ($(DEB_HOST_GNU_SYSTEM),linux-gnu)
+UBINFILES += sys-utils/ipcs misc-utils/setterm
+endif
 SBINFILES = disk-utils/mkswap
 
 ifeq ($(DEB_HOST_GNU_SYSTEM),linux-gnu)
@@ -79,11 +83,15 @@
 SBINFILES += hwclock/hwclock
 endif
 BINFILES  += sys-utils/dmesg
-SBINFILES += disk-utils/{blockdev,raw} mount/pivot_root login-utils/agetty
+SBINFILES += disk-utils/{blockdev,raw} mount/pivot_root
 UBINFILES += disk-utils/fdformat
 USBINFILES = sys-utils/readprofile disk-utils/elvtune # disk-utils/setfdprm
 endif
 
+ifneq ($(DEB_HOST_GNU_SYSTEM),gnu)
+SBINFILES += login-utils/agetty
+endif
+
 UBINFILES2= misc-utils/chkdupexe # debian/fdformat # don't strip these
 
 ifeq ($(DEB_HOST_GNU_SYSTEM),linux-gnu)
@@ -111,11 +119,14 @@
 MAN8FILES += hwclock/hwclock.8
 endif
 MAN1FILES += sys-utils/readprofile.1
-MAN8FILES += login-utils/agetty.8 disk-utils/{blockdev.8,elvtune.8} \
+MAN8FILES += disk-utils/{blockdev.8,elvtune.8} \
 	     sys-utils/dmesg.8 mount/pivot_root.8 \
 	     disk-utils/fdformat.8 disk-utils/raw.8 # disk-utils/setfdprm.8
 endif
 
+ifneq ($(DEB_HOST_GNU_SYSTEM),gnu)
+MAN8FILES += login-utils/agetty.8
+endif
 
 EXAMPLES = fdisk/sfdisk.examples getopt/getopt-{test,parse}.{ba,tc}sh
 ifeq ($(DEB_HOST_GNU_SYSTEM),linux-gnu)
@@ -123,8 +134,11 @@
 endif
 
 INFOFILES = sys-utils/ipc.info
-ifeq ($(DEB_HOST_GNU_SYSTEM),linux-gnu)
+ifneq ($(DEB_HOST_GNU_SYSTEM),gnu)
 DOCFILES  = login-utils/{README.getty,README.modems-with-agetty,README.poeigl}
+endif
+
+ifeq ($(DEB_HOST_GNU_SYSTEM),linux-gnu)
 ifneq ($(arch),$(nohwclock))
 DOCFILES += debian/README.Debian.hwclock
 endif
@@ -204,11 +218,11 @@
 	mv -f debian/tmp/sbin/fdisk debian/tmp/sbin/ddisk
 	mv -f debian/tmp/usr/share/man/man8/fdisk.8 debian/tmp/usr/share/man/man8/ddisk.8
 endif
-ifeq ($(DEB_HOST_GNU_SYSTEM),gnu)
+ifneq ($(DEB_HOST_GNU_SYSTEM),linux-gnu)
 	(cd debian/tmp/sbin ; mv mkswap mkswap.linux)
 	(cd debian/tmp/usr/share/man/man8 ; mv mkswap.8 mkswap.linux.8)
 endif
-ifeq ($(DEB_HOST_GNU_SYSTEM),linux-gnu)
+ifneq ($(DEB_HOST_GNU_SYSTEM),gnu)
 	(cd debian/tmp/sbin ; mv agetty getty)
 	(cd debian/tmp/usr/share/man/man8 ; mv agetty.8 getty.8)
 ifneq ($(arch), $(nohwclock))
diff -ur util-linux-2.12p.old/disk-utils/Makefile util-linux-2.12p/disk-utils/Makefile
--- util-linux-2.12p.old/disk-utils/Makefile	2004-12-21 18:14:16.000000000 +0100
+++ util-linux-2.12p/disk-utils/Makefile	2005-10-06 19:17:28.000000000 +0200
@@ -16,16 +16,19 @@
 
 SBIN= 	mkfs mkswap blockdev elvtune fsck.minix mkfs.minix mkfs.bfs
 
-USRBIN=	fdformat isosize
+USRBIN=	isosize
 
-ETC=	fdprm
-
-MAYBE=  setfdprm raw fsck.cramfs mkfs.cramfs
+MAYBE=  raw fsck.cramfs mkfs.cramfs
 
+ifeq ($(shell uname -s), Linux)
+USRBIN += fdformat
+ETC=	fdprm
+MAYBE += setfdprm
 ifneq "$(HAVE_FDUTILS)" "yes"
 USRBIN:=$(USRBIN) setfdprm
 MAN8:=$(MAN8) setfdprm.8
 endif
+endif
 
 ifeq "$(HAVE_RAW_H)" "yes"
 USRBIN:=$(USRBIN) raw
diff -ur util-linux-2.12p.old/disk-utils/fsck.cramfs.c util-linux-2.12p/disk-utils/fsck.cramfs.c
--- util-linux-2.12p.old/disk-utils/fsck.cramfs.c	2005-10-06 18:45:25.000000000 +0200
+++ util-linux-2.12p/disk-utils/fsck.cramfs.c	2005-10-06 19:13:12.000000000 +0200
@@ -76,7 +76,11 @@
 
 #define PAD_SIZE 512
 
+#if defined(__linux__)
 #include <asm/page.h>
+#elif defined(__FreeBSD_kernel__)
+#include <machine/param.h>
+#endif
 #ifdef PAGE_SIZE
 #define PAGE_CACHE_SIZE ((int) PAGE_SIZE)
 #elif defined __ia64__
diff -ur util-linux-2.12p.old/login-utils/agetty.c util-linux-2.12p/login-utils/agetty.c
--- util-linux-2.12p.old/login-utils/agetty.c	2002-07-29 09:36:42.000000000 +0200
+++ util-linux-2.12p/login-utils/agetty.c	2005-10-06 19:57:13.000000000 +0200
@@ -18,7 +18,12 @@
 #include <stdlib.h>
 #include <string.h>
 #include <sys/ioctl.h>
+#ifdef __linux__
 #include <termio.h>
+#else
+#include <termios.h>
+#define termio termios
+#endif
 #include <signal.h>
 #include <errno.h>
 #include <sys/types.h>
@@ -33,11 +38,9 @@
 #include "xstrncpy.h"
 #include "nls.h"
 
-#ifdef __linux__
 #include "pathnames.h"
 #include <sys/param.h>
 #define USE_SYSLOG
-#endif
 
  /* If USE_SYSLOG is undefined all diagnostics go directly to /dev/console. */
 
@@ -113,6 +116,11 @@
 #define	TCSETAW	TCSETSW
 #endif
 
+/* Linux-specific */
+#ifndef CBAUD
+#define CBAUD 0
+#endif
+
  /*
   * This program tries to not use the standard-i/o library.  This keeps the
   * executable small on systems that do not have shared libraries (System V
@@ -299,9 +307,7 @@
 
     parse_args(argc, argv, &options);
 
-#ifdef __linux__
 	setsid();
-#endif
 	
     /* Update the utmp file. */
 
@@ -686,8 +692,13 @@
      * 5 seconds seems to be a good value.
      */
 
+#ifdef __linux__
     if (ioctl(0, TCGETA, tp) < 0)
 	error("%s: ioctl: %m", tty);
+#else
+    if (tcgetattr (0, tp) < 0)
+	error("%s: tcgetattr: %m", tty);
+#endif
 
     /*
      * It seems to be a terminal. Set proper protections and ownership. Mode
@@ -723,9 +734,11 @@
      * reads will be done in raw mode anyway. Errors will be dealt with
      * lateron.
      */
-#ifdef __linux__
     /* flush input and output queues, important for modems! */
+#ifdef __linux__
     (void) ioctl(0, TCFLSH, TCIOFLUSH);
+#else
+    tcflush (0, TCIOFLUSH);
 #endif
 
     tp->c_cflag = CS8 | HUPCL | CREAD | speed;
@@ -733,7 +746,10 @@
 	tp->c_cflag |= CLOCAL;
     }
 
-    tp->c_iflag = tp->c_lflag = tp->c_oflag = tp->c_line = 0;
+    tp->c_iflag = tp->c_lflag = tp->c_oflag = 0;
+#ifdef __linux__
+    tp->c_line = 0;
+#endif
     tp->c_cc[VMIN] = 1;
     tp->c_cc[VTIME] = 0;
 
@@ -744,7 +760,11 @@
 	tp->c_cflag |= CRTSCTS;
 #endif
 
+#ifdef __linux__
     (void) ioctl(0, TCSETA, tp);
+#else
+    tcsetattr (0, TCSANOW, tp);
+#endif
 
     /* go to blocking input even in local mode */
     fcntl(0, F_SETFL, fcntl(0, F_GETFL, 0) & ~O_NONBLOCK);
@@ -764,6 +784,11 @@
     char   *bp;
     int     nread;
 
+#ifdef __FreeBSD_kernel__
+    ioctl (0, TIOCSCTTY);
+    putenv ("TERM=cons25");
+#endif
+
     /*
      * This works only if the modem produces its status code AFTER raising
      * the DCD line, and if the computer is fast enough to set the proper
@@ -788,7 +813,11 @@
     tp->c_iflag |= ISTRIP;			/* enable 8th-bit stripping */
     vmin = tp->c_cc[VMIN];
     tp->c_cc[VMIN] = 0;				/* don't block if queue empty */
+#ifdef __linux__
     (void) ioctl(0, TCSETA, tp);
+#else
+    tcsetattr (0, TCSANOW, tp);
+#endif
 
     /*
      * Wait for a while, then read everything the modem has said so far and
@@ -812,7 +841,11 @@
 
     tp->c_iflag = iflag;
     tp->c_cc[VMIN] = vmin;
+#if defined(__linux__) && !defined(__GLIBC__)
     (void) ioctl(0, TCSETA, tp);
+#else
+    tcsetattr (0, TCSANOW, tp);
+#endif
 }
 
 /* do_prompt - show login prompt, optionally preceded by /etc/issue contents */
@@ -835,8 +868,11 @@
     if ((op->flags & F_ISSUE) && (fd = fopen(op->issue, "r"))) {
 	oflag = tp->c_oflag;			/* save current setting */
 	tp->c_oflag |= (ONLCR | OPOST);		/* map NL in output to CR-NL */
+#ifdef __linux__
 	(void) ioctl(0, TCSETAW, tp);
-
+#else
+        tcsetattr (0, TCSADRAIN, tp);
+#endif
 
 	while ((c = getc(fd)) != EOF)
 	{
@@ -947,18 +983,22 @@
 	fflush(stdout);
 
 	tp->c_oflag = oflag;			/* restore settings */
+#ifdef __linux__
 	(void) ioctl(0, TCSETAW, tp);		/* wait till output is gone */
+#else
+        tcsetattr (0, TCSADRAIN, tp);
+#endif
 	(void) fclose(fd);
     }
 #endif
-#ifdef __linux__
+#ifdef MAXHOSTNAMELEN
 	{
 		char hn[MAXHOSTNAMELEN+1];
 
 		(void) gethostname(hn, MAXHOSTNAMELEN);
 		write(1, hn, strlen(hn));
 	}
-#endif		
+#endif
     (void) write(1, LOGIN, sizeof(LOGIN) - 1);	/* always show login prompt */
 }
 
@@ -973,7 +1013,11 @@
     baud_index = (baud_index + 1) % op->numspeed;
     tp->c_cflag &= ~CBAUD;
     tp->c_cflag |= op->speeds[baud_index];
+#ifdef __linux__
     (void) ioctl(0, TCSETA, tp);
+#else
+    tcsetattr (0, TCSANOW, tp);
+#endif
 }
 
 /* get_logname - get user name, establish parity, speed, erase, kill, eol */
@@ -1003,7 +1047,11 @@
     /* Flush pending input (esp. after parsing or switching the baud rate). */
 
     (void) sleep(1);
+#ifdef __linux__
     (void) ioctl(0, TCFLSH, TCIFLUSH);
+#else
+    tcflush (0, TCIFLUSH);
+#endif
 
     /* Prompt for and read a login name. */
 
@@ -1105,9 +1153,9 @@
     tp->c_cc[VQUIT] = DEF_QUIT;			/* default quit */
     tp->c_cc[VEOF] = DEF_EOF;			/* default EOF character */
     tp->c_cc[VEOL] = DEF_EOL;
-#ifdef __linux__
+#if defined(__linux__)
     tp->c_cc[VSWTC] = DEF_SWITCH;		/* default switch character */
-#else
+#elif defined(VSWTCH)
     tp->c_cc[VSWTCH] = DEF_SWITCH;		/* default switch character */
 #endif
 
@@ -1140,9 +1188,15 @@
     /* Account for upper case without lower case. */
 
     if (cp->capslock) {
+#ifdef IUCLC
 	tp->c_iflag |= IUCLC;
+#endif
+#ifdef XCASE
 	tp->c_lflag |= XCASE;
+#endif
+#ifdef OLCUC
 	tp->c_oflag |= OLCUC;
+#endif
     }
     /* Optionally enable hardware flow control */
 
@@ -1153,8 +1207,13 @@
 
     /* Finally, make the new settings effective */
 
+#ifdef __linux__
     if (ioctl(0, TCSETA, tp) < 0)
 	error("%s: ioctl: TCSETA: %m", op->tty);
+#else
+    if (tcsetattr (0, TCSANOW, tp) < 0)
+	error("%s: tcsetattr: %m", op->tty);
+#endif
 }
 
 /* caps_lock - string contains upper case without lower case */
diff -ur util-linux-2.12p.old/misc-utils/Makefile util-linux-2.12p/misc-utils/Makefile
--- util-linux-2.12p.old/misc-utils/Makefile	2004-12-05 20:09:12.000000000 +0100
+++ util-linux-2.12p/misc-utils/Makefile	2005-10-06 19:37:23.000000000 +0200
@@ -22,7 +22,11 @@
 USRBIN=		cal chkdupexe ddate logger look mcookie \
 		namei rename script whereis
 
-MAYBE=		reset setterm
+MAYBE=		reset
+
+ifeq ($(shell uname -s), Linux)
+MAYBE += setterm
+endif
 
 ifeq "$(HAVE_RESET)" "no"
 USRBIN:=$(USRBIN) reset
@@ -39,10 +43,12 @@
 MAN1:=$(MAN1) kill.1
 endif
 
+ifeq ($(shell uname -s), Linux)
 ifeq "$(HAVE_NCURSES)" "yes"
 USRBIN:=$(USRBIN) setterm
 MAN1:=$(MAN1) setterm.1
 endif
+endif
 
 # For script only
 LIBPTY=
@@ -51,7 +57,9 @@
 endif
 
 # Programs requiring special compilation
+ifeq ($(shell uname -s), Linux)
 NEEDS_CURSES=  setterm
+endif
 NEEDS_OPENPTY= script
 
 all: $(BIN) $(USRBIN) $(USRBIN.NONSHADOW)
@@ -99,9 +107,11 @@
 write.o: $(LIB)/carefulputc.h
 write: write.o $(LIB)/carefulputc.o
 
+ifeq ($(shell uname -s), Linux)
 ifeq "$(HAVE_NCURSES)" "yes"
 setterm: setterm.o
 endif
+endif
 
 install: all
 	$(INSTALLDIR) $(BINDIR) $(USRBINDIR)
diff -ur util-linux-2.12p.old/misc-utils/mcookie.c util-linux-2.12p/misc-utils/mcookie.c
--- util-linux-2.12p.old/misc-utils/mcookie.c	2002-03-09 00:00:52.000000000 +0100
+++ util-linux-2.12p/misc-utils/mcookie.c	2005-10-06 19:34:11.000000000 +0200
@@ -20,7 +20,7 @@
  *
  */
 
-#ifdef __linux__
+#if defined(__linux__) || defined(__GLIBC__)
 #define HAVE_GETTIMEOFDAY 1
 #endif
 
diff -ur util-linux-2.12p.old/misc-utils/script.c util-linux-2.12p/misc-utils/script.c
--- util-linux-2.12p.old/misc-utils/script.c	2004-03-26 18:07:16.000000000 +0100
+++ util-linux-2.12p/misc-utils/script.c	2005-10-06 19:34:51.000000000 +0200
@@ -54,10 +54,8 @@
 #include <sys/signal.h>
 #include "nls.h"
 
-#ifdef __linux__
 #include <unistd.h>
 #include <string.h>
-#endif
 
 #include "../defines.h"
 #ifdef HAVE_openpty
diff -ur util-linux-2.12p.old/sys-utils/Makefile util-linux-2.12p/sys-utils/Makefile
--- util-linux-2.12p.old/sys-utils/Makefile	2004-11-15 18:47:47.000000000 +0100
+++ util-linux-2.12p/sys-utils/Makefile	2005-10-06 19:42:31.000000000 +0200
@@ -17,9 +17,9 @@
 # Where to put binaries?
 # See the "install" rule for the links. . .
 
-BIN=            arch dmesg
+BIN=            arch
 
-USRBIN=		cytune flock ipcrm ipcs renice setsid
+USRBIN=		flock ipcrm renice setsid
 
 USRSBIN=	readprofile tunelp
 
@@ -27,6 +27,11 @@
 
 NOTMADE=
 
+ifeq ($(shell uname -s), Linux)
+BIN += dmesg
+USRBIN += cytune ipcs
+endif
+
 CWFLAGS := $(subst -Wmissing-prototypes,,$(CFLAGS))
 
 ifeq "$(HAVE_SLN)" "no"
diff -ur util-linux-2.12p.old/text-utils/more.c util-linux-2.12p/text-utils/more.c
--- util-linux-2.12p.old/text-utils/more.c	2004-12-22 11:46:25.000000000 +0100
+++ util-linux-2.12p/text-utils/more.c	2005-10-06 19:45:17.000000000 +0200
@@ -76,6 +76,15 @@
 
 #define stty(fd,argp)  tcsetattr(fd,TCSANOW,argp)
 
+/* Linux-specific */
+#ifndef CBAUD
+#define CBAUD 0
+#endif
+
+#ifndef TABDLY
+#define TABDLY 0
+#endif
+
 /* some function declarations */
 void initterm(void);
 void kill_line(void);
@@ -1560,10 +1569,6 @@
     }
     if (feof (file)) {
 	if (!no_intty) {
-#ifndef __linux__
-				/* No longer in libc 4.5.8. . . */
-	    file->_flags &= ~STDIO_S_EOF_SEEN; /* why doesn't fseek do this ??!!??! */
-#endif
 	    Currline = saveln;
 	    Fseek (file, startline);
 	}
